### PR TITLE
planner: add the implementation for SetCollationExpr

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
A new `SetCollationExpr` is added to support syntax like:

```
tidb> select weight_string('a' collate utf8mb4_general_ci);
```

In which the clause `collate utf8mb4_general_ci` was ignored by TiDB before this PR.

Corresponding parser PR: https://github.com/pingcap/parser/pull/749

### What is changed and how it works?
Set the collation of the child of this expression to the one specified by it.

### Check List 

Tests 

 - Unit test
 
Code changes

 - NA

Side effects

 - NA

Related changes

 - NA

Release note

 - NA

### CODE REVIEW

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -982,6 +984,19 @@ func (er *expressionRewriter) Leave(originInNode ast.Node) (retNode ast.Node, ok
 			Value:   types.NewStringDatum(v.Selector.String()),
 			RetType: types.NewFieldType(mysql.TypeVarchar),
 		}, types.EmptyName)
+	case *ast.SetCollationExpr:
+		arg := er.ctxStack[len(er.ctxStack)-1]
+		// SetCollationExpr sets the collation explicitly, even when the evaluation type of the expression is non-string.
+		if _, ok := arg.(*expression.Column); ok {
```
Comment (AilinKid) [inline planner/core/expression_rewriter.go]: any test about the column arg?

File: planner/core/expression_rewriter.go [hunk 2]
```diff
@@ -982,6 +984,19 @@ func (er *expressionRewriter) Leave(originInNode ast.Node) (retNode ast.Node, ok
...
+			// Wrap a cast here to avoid changing the original FieldType of the column expression.
+			casted := expression.WrapWithCastAsString(er.sctx, arg)
+			casted.GetType().Collate = v.Collate
+			er.ctxStackAppend(casted, types.EmptyName)
+		} else {
+			// For constant and scalar function, we can set its collate directly.
+			arg.GetType().Collate = v.Collate
```
Comment (wjhuang2016) [inline planner/core/expression_rewriter.go]: We need to check if the collation is compatible with charset

File: planner/core/expression_rewriter.go [hunk 3]
```diff
@@ -982,6 +984,19 @@ func (er *expressionRewriter) Leave(originInNode ast.Node) (retNode ast.Node, ok
 			Value:   types.NewStringDatum(v.Selector.String()),
 			RetType: types.NewFieldType(mysql.TypeVarchar),
 		}, types.EmptyName)
+	case *ast.SetCollationExpr:
+		arg := er.ctxStack[len(er.ctxStack)-1]
+		// SetCollationExpr sets the collation explicitly, even when the evaluation type of the expression is non-string.
+		if _, ok := arg.(*expression.Column); ok {
+			// Wrap a cast here to avoid changing the original FieldType of the column expression.
+			casted := expression.WrapWithCastAsString(er.sctx, arg)
```
Comment (wjhuang2016) [inline planner/core/expression_rewriter.go]: Wy need to avoid it? Could you give an example?

File: planner/core/expression_rewriter.go (hunk 1 above)
Comment (bb7133) [reply planner/core/expression_rewriter.go]: added, thanks.

File: planner/core/expression_rewriter.go (hunk 3 above)
Comment (bb7133) [reply planner/core/expression_rewriter.go]: If we have `select a, a collate utf8_bin from t`, and we set `utf8_bin` for `a` directly, we may have incorrect result for `select a`.

File: go.sum [hunk 4]
```diff
@@ -93,8 +93,8 @@ github.com/golang/protobuf v0.0.0-20180814211427-aa810b61a9c7/go.mod h1:6lQm79b+
 github.com/golang/protobuf v1.2.0/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
 github.com/golang/protobuf v1.3.1/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
 github.com/golang/protobuf v1.3.2/go.mod h1:6lQm79b+lXiMfvg/cZm0SGofjICqVBUtrP5yJMmIC1U=
-github.com/golang/protobuf v1.3.3 h1:gyjaxf+svBWX08ZjK86iN9geUJF0H6gp2IRKX6Nf6/I=
-github.com/golang/protobuf v1.3.3/go.mod h1:vzj43D7+SQXF/4pzW/hwtAqwc6iTitCiVSaWz5lYuqw=
+github.com/golang/protobuf v1.3.4 h1:87PNWwrRvUSnqS4dlcBU/ftvOIBep4sYuBLlh6rX2wk=
+github.com/golang/protobuf v1.3.4/go.mod h1:vzj43D7+SQXF/4pzW/hwtAqwc6iTitCiVSaWz5lYuqw=
```
Comment (djshow832) [inline go.sum]: Other repos are also updated, is that expected?

Comment (bb7133) [reply go.sum]: It is updated by go mod automatically. I think that's fine.

File: planner/core/expression_rewriter.go (hunk 2 above)
Comment (bb7133) [reply planner/core/expression_rewriter.go]: Done

**Pull request merged**